packages/query/src/block-processor.ts

```diff
@@ -169,16 +170,17 @@ export class BlockProcessor implements BlockProcessorInterface {
         await this.indexedDb.saveFmdParams(compactBlock.fmdParameters);
       }
       if (compactBlock.gasPrices) {
-        await this.indexedDb.saveGasPrices(
-          new GasPrices({
-            assetId: this.stakingAssetId,
-            ...compactBlock.gasPrices,
-          }),
-        );
+        await this.indexedDb.saveGasPrices({
+          ...toPlainMessage(compactBlock.gasPrices),
+          assetId: toPlainMessage(this.stakingAssetId),
+        });
       }
       if (compactBlock.altGasPrices.length) {
-        for (const gasPrice of compactBlock.altGasPrices) {
-          await this.indexedDb.saveGasPrices(gasPrice);
+        for (const altGas of compactBlock.altGasPrices) {
+          await this.indexedDb.saveGasPrices({
+            ...toPlainMessage(altGas),
+            assetId: getAssetIdFromGasPrices(altGas),
+          });
```

Comment:
issue: can we just put the guard inside storage so callers don't have to worry about this?

```ts
  async saveGasPrices(value: PartialMessage<GasPrices>): Promise<void> {
    if (!value.assetId) {
      value.assetId = this.stakingTokenAssetId;
    }
    
    await this.u.update({
      table: 'GAS_PRICES',
      value: new GasPrices(value).toJson() as Jsonified<GasPrices>,
    });
  }
```

This also doesn't require `toPlainMessage` which isn't entirely obvious from a glance how it serializes all the fields.

Reply:
Actually, scratch that. We have to handle Um/alts differently. Something like this would work:

```ts
// used in main block processor loop
await this.saveGasPrices(compactBlock);

=====
// additional method

  private async saveGasPrices(block: CompactBlock) {
    const pricesToSave = [...block.altGasPrices];

    // Undefined field implies UM. Manually overriding as this is commonly not passed.
    if (block.gasPrices) {
      pricesToSave.push(
        new GasPrices({
          ...block.gasPrices,
          assetId: this.stakingAssetId,
        }),
      );
    }

    for (const price of pricesToSave) {
      if (!price.assetId) {
        console.warn(
          `BlockHeight: ${block.height}. No asset id for gas prices: ${JSON.stringify(price.toJson())}`,
        );
        continue;
      }
      await this.indexedDb.saveGasPrices(price);
    }
  }
```

Reply:
callers should be explicit in intent. a well-typed parameter means callers don't have to worry - the type system will protect them.

unanticipated fallback behavior upon absent input (not default input, but literally absent input) is a greater source of concern. 

plainmessage does not serialize fields. it represents the message type as a plain object. it is the same as the class instance, sans class instance methods.

it is the simplest way to express that a parameter requires data present for every field of the message type. using `Required` on the class would require non-message input, using the class led us to the present situation, and the third option is just a literal redefinition of what the plainmessage type provides.

any class instance or object literal meeting the message type satisfies it.

this information is available from the docstring on the type and on the function. its existence is explained and its use in specific cases such as this are recommended in bufbuild/connectrpc documentation.